### PR TITLE
Implement bindings to the History API

### PR DESCRIPTION
**Pull request by tomhoule**
They work in the small example included in the PR. There are however a few questions left open:

- I don't think the unit test for `PopStateEvent` is run, it never fails.
- Is the signature for `History::state` the preferred way to extract typed values from js?
- Should the example stay there? It might gain from being a bit less trivial in that case.
- Is the code formatted correctly?

Looking forward to your feedback :)

### CODE REVIEW

File: src/webapi/history.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+reference_boilerplate! {
+    History,
+    instanceof History
+}
+
+impl History {
+    /// `History.pushState` allows you to add history entries.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/History_API#The_pushState%28%29_method)
+    pub fn push_state<T: JsSerializable>(&self, state: T, title: Option<&str>, url: String) {
```
Comment (koute) [inline src/webapi/history.rs]: In the [specs](https://html.spec.whatwg.org/multipage/history.html#the-history-interface) I can see that the `title` is marked as being non-optional (is not inside square brackets) while the `url` is, so shouldn't the prototype be `&self, state: T, title: &str, url: Option< &str >`?

File: src/webapi/history.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/History)
+pub struct History(Reference);
+
+reference_boilerplate! {
+    History,
+    instanceof History
+}
+
+impl History {
+    /// `History.pushState` allows you to add history entries.
```
Comment (koute) [inline src/webapi/history.rs]: In general we don't reference the original JS `camelCased` name in the docs, and there is no point in referencing the name of the function you're documenting anyway. (:

**Comment by koute** [comment]
> I don't think the unit test for PopStateEvent is run, it never fails.

Yes, it's not run. You need to delete the `[cfg(...)]` there and run the test without passing `--nodejs` to `cargo-web` (it will then run the test in a web browser, which is kinda hack-ish and experimental for now, which is why it isn't yet done by default)

> Should the example stay there? It might gain from being a bit less trivial in that case.

I don't mind extra examples. (: 

> Is the code formatted correctly?

Don't worry about it; worrying about formatting is not something humans should waste time on. I'll have to run the whole project through `rustfmt` sooner or later.

**Comment by tomhoule** [comment]
Just a quick message to say my laptop died so it will be a few days until I replace it and amend the PR :)

**Comment by koute** [comment]
@tomhoule Sure; thanks for the heads up. (:

**Comment by tomhoule** [comment]
I managed to free up some time to amend the PR. I think I addressed everything from the last review, managed to run the tests, but I haven't improved the example. It just does some light history manipulation and logs to the console.

Maybe rendering history state to the DOM would make it more convincing? I'll try to think of something better over the weekend :)

File: src/webapi/history.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            @{self}.go(@{offset})
+        };
+    }
+
+    /// Move one step backward through history.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/History_API#Traveling_through_history)
+    pub fn back(&self) {
+        js! { @(no_return)
+            @{self}.back()
```
Comment (koute) [inline src/webapi/history.rs]: Missing a semicolon. (:

File: src/webapi/history.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        };
+    }
+
+    /// Move one step forward through history.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/History_API#Traveling_through_history)
+    pub fn forward(&self) {
+        js! { @(no_return)
+            @{self}.forward()
```
Comment (koute) [inline src/webapi/history.rs]: Missing a semicolon here too.

File: src/webapi/history.rs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    /// You can use the go() method to load a specific page from session history, identified by its
+    /// relative position to the current page (with the current page being, of course, relative
+    /// index 0).
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/History_API#Traveling_through_history)
+    pub fn go(&self, offset: i32) {
+        js! { @(no_return)
+            @{self}.go(@{offset})
```
Comment (koute) [inline src/webapi/history.rs]: Missing semicolon here.

File: src/webapi/history.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
+#![cfg(feature = "serde")]
```
Comment (koute) [inline src/webapi/history.rs]: Looks like you forgot to remove this? (:

File: src/webapi/history.rs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    pub fn forward(&self) {
+        js! { @(no_return)
+            @{self}.forward()
+        };
+    }
+
+    /// Returns the current number of history entries.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/History)
+    pub fn len(&self) -> i32 {
```
Comment (koute) [inline src/webapi/history.rs]: `i32` -> `usize` to keep with Rust`s convention

File: src/webapi/history.rs (hunk 4 above)
Comment (tomhoule) [reply src/webapi/history.rs]: My habit of writing JS without semicolons is showing :) Fixed!

File: src/webapi/history.rs (hunk 6 above)
Comment (tomhoule) [reply src/webapi/history.rs]: Oh, indeed, my bad.

**Comment by tomhoule** [comment]
`usize` does not work as a return type for `history.len()`, there is apparently a missing trait implementation. When trying to compile the example I get the following error:

```
error[E0277]: the trait bound `stdweb::private::Newtype<_, usize>: stdweb::private::JsSerializableOwned` is not satisfied
  --> src/main.rs:53:9
   |
53 |         js!(console.log("number of history entries: ", @{history_length}));
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `stdweb::private::JsSerializableOwned` is not implemented for `stdweb::private::Newtype<_, usize>`
   |
   = help: the following implementations were found:
             <stdweb::private::Newtype<(stdweb::webcore::serialization::FunctionTag, (A4, A5, A6, A7, A8, A9, A10, A11, A12)), F> as stdweb::private::JsSerializableOwned>
             <stdweb::private::Newtype<(stdweb::webcore::serialization::FunctionTag, (A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12)), F> as stdweb::private::JsSerializableOwned>
             <stdweb::private::Newtype<(stdweb::webcore::serialization::FunctionTag, (A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12)), std::option::Option<F>> as stdweb::private::JsSerializableOwned>
             <stdweb::private::Newtype<(stdweb::webcore::serialization::FunctionTag, (A10, A11, A12)), std::option::Option<F>> as stdweb::private::JsSerializableOwned>
           and 39 others
   = note: required by `stdweb::private::JsSerializableOwned::into_js_owned`
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

It could probably be added in `webcore/serialization.rs` if my understanding is correct. I don't see implementations for `u64` there either. wasm is 32 bits only so far if I remember correctly.

Is it something that should be implemented in this PR?

**Comment by koute** [comment]
> It could probably be added in webcore/serialization.rs if my understanding is correct. I don't see implementations for u64 there either.

Hmm... well, yes, the problem with passing 64-bit numbers into the JavaScript is that the maximum number you can safely pass is 2^53 - 1, which is why neither the `u64` conversion is implemented nor `usize` (as *technically* `usize` could be 64 bit in the future).

I guess for pragmatic reasons we *could* make `usize` serializable (since it's safe on 32-bit targets), and when the 64-bit WASM target is a thing (**if** it's going to be a thing at all) add a check when serializing whenever the number is less than 2^53.

Feel free to add it! Just please put it in a separate commit from the `History` changes.

**Comment by Pauan** [comment]
`History.prototype.length` (and almost all lengths in JavaScript, including `Array.prototype.length`) returns `u32`:

https://html.spec.whatwg.org/multipage/history.html#the-history-interface

So using `usize` is wrong in any case, regardless of whether wasm supports 64-bit or not.

**Comment by koute** [comment]
Well, yes, but, conversely any length that JavaScript will give us will be always convertible to an `usize`; the problem is only when going the other way.

If I think about this really hard I guess you're right - `usize` in Rust is explicitly defined as to be as big as a pointer, and those sizes have absolutely nothing to do with Rust's pointer size! Even though it's more convenient to have those as `usize`s since then you can interop with Rust's `Vec` and others without any casts.

I'll merge this as-is and later I'll fix it myself along with other instances where we currently return `usize`.

@tomhoule Thanks for the PR!

@Pauan Thanks for keeping an eye out!

**Comment by Pauan** [comment]
@koute Yeah, needing to cast `u32 <=> usize` sucks, but it is technically correct.

If `usize` auto-casted to / from `u32` then we could use `usize`, but that feels wrong, since it would either wrap around (bad) or truncate (also bad).

Because casting causes a loss of information, I feel like it should be manual, not automatic, and therefore the correct type is `u32`

**Pull request merged**